### PR TITLE
Uar 431 ceased date field and validation roe api

### PR DESCRIPTION
**Pull request by cdrainCH**
### JIRA link

https://companieshouse.atlassian.net/jira/software/c/projects/UAR/boards/481?modal=detail&selectedIssue=UAR-431&sprint=2781

### Change description

Adding ceased date field to BOs in API, testing and validating as needed

### Work checklist

- [ ] Tests added where applicable

### Merge instructions

We are committed to keeping commit history clean, consistent and linear. To achieve this, this commit should be structured as follows:

```
<type>[optional scope]: <description>
```

and contain the following structural elements:

- fix: a commit that patches a bug in your codebase (this correlates with PATCH in semantic versioning),
- feat: a commit that introduces a new feature to the codebase (this correlates with MINOR in semantic versioning),
- BREAKING CHANGE: a commit that has a footer `BREAKING CHANGE:` introduces a breaking API change (correlating with MAJOR in semantic versioning). A BREAKING CHANGE can be part of commits of any type,
- types other than `fix:` and `feat:` are allowed, for example `build:`, `chore:`, `ci:`, `docs:`, `style:`, `refactor:`, `perf:`, `test:`, and others,
- footers other than `BREAKING CHANGE: <description>` may be provided.

### CODE REVIEW

File: src/main/java/uk/gov/companieshouse/overseasentitiesapi/validation/BeneficialOwnerIndividualValidator.java [hunk 1]
```diff
@@ -154,4 +158,10 @@ private boolean validateNatureOfControl(List<NatureOfControlType> fields, Errors
         String qualifiedFieldName = getQualifiedFieldName(OverseasEntitySubmissionDto.BENEFICIAL_OWNERS_INDIVIDUAL_FIELD, NATURE_OF_CONTROL_FIELDS);
         return NatureOfControlValidators.checkAtLeastOneFieldHasValue(fields, qualifiedFieldName, errors, loggingContext);
     }
+
+    private boolean validateCeasedDate(LocalDate ceasedDate, LocalDate startDate, Errors errors, String loggingContext) {
+        String qualifiedFieldName = getQualifiedFieldName(OverseasEntitySubmissionDto.BENEFICIAL_OWNERS_INDIVIDUAL_FIELD, BeneficialOwnerIndividualDto.CEASED_DATE_FIELD);
+        return DateValidators.isDateInPast(ceasedDate, qualifiedFieldName, errors, loggingContext)
```
Comment (AndreKramer7654) [inline src/main/java/uk/gov/companieshouse/overseasentitiesapi/validation/BeneficialOwnerIndividualValidator.java:164]: Is today included? Should be valid to say ceased today.

File: src/main/java/uk/gov/companieshouse/overseasentitiesapi/validation/BeneficialOwnerIndividualValidator.java [hunk 2]
```diff
@@ -154,4 +158,10 @@ private boolean validateNatureOfControl(List<NatureOfControlType> fields, Errors
...
+                && DateValidators.isCeasedDateAfterStartDate(ceasedDate, startDate, qualifiedFieldName, errors, loggingContext);
```
Comment (AndreKramer7654) [inline src/main/java/uk/gov/companieshouse/overseasentitiesapi/validation/BeneficialOwnerIndividualValidator.java:165]: Soon we will have a filing date. Need validation against that too. Probably need to put something on the backlog for that.

File: src/main/java/uk/gov/companieshouse/overseasentitiesapi/validation/utils/DateValidators.java [hunk 3]
```diff
@@ -31,4 +31,14 @@ public static boolean isDateWithinLast3Months(LocalDate compareToDate, String qu
         ApiLogger.infoContext(loggingContext, qualifiedFieldName + " must be in the past 3 months");
         return false;
     }
+
+    public static boolean isCeasedDateAfterStartDate(LocalDate ceasedDate, LocalDate startDate, String qualifiedFieldName, Errors errors, String loggingContext) {
```
Comment (AndreKramer7654) [inline src/main/java/uk/gov/companieshouse/overseasentitiesapi/validation/utils/DateValidators.java:35]: what if they are equal? Not before does not always mean after. Not sure they can cease on appointed date - probably someone trying to "remove" a BO so important to get that right.

Comment (cdrainCH) [reply src/main/java/uk/gov/companieshouse/overseasentitiesapi/validation/utils/DateValidators.java:35]: Good point. Method now checks that ceased date is neither before nor the same as start date

File: src/main/java/uk/gov/companieshouse/overseasentitiesapi/validation/utils/ValidationMessages.java [hunk 4]
```diff
@@ -19,4 +19,5 @@ private ValidationMessages() {}
     public static final String NATIONALITY_NOT_ON_LIST_ERROR_MESSAGE = "%s is not on the list of nationalities";
     public static final String SECOND_NATIONALITY_SHOULD_BE_DIFFERENT = "%s should not be the same as the nationality given";
     public static final String DUPLICATE_TRUST_ID = "Duplicate Trust Id for %s";
+    public static final String CEASED_DATE_BEFORE_START_DATE_ERROR_MESSAGE = "%s must be in the past";
```
[the remark this answers is not in the document]
Comment (AndreKramer7654) [reply src/main/java/uk/gov/companieshouse/overseasentitiesapi/validation/utils/ValidationMessages.java:22]: We have yet to do the web side so for now ceased_date will be null.

[the remark this answers is not in the document]
Comment (cdrainCH) [reply src/main/java/uk/gov/companieshouse/overseasentitiesapi/validation/utils/ValidationMessages.java:22]: @mwestacott I opened a new PR addressing the error messages. Also I believe the method name above should be fine, as we are working with the ceased date having to be after the appointed date and not before or on the appointed date. Here is the link for the new PR https://github.com/companieshouse/overseas-entities-api/pull/198

[the remark this answers is not in the document]
Comment (mwestacott) [reply src/main/java/uk/gov/companieshouse/overseasentitiesapi/validation/utils/ValidationMessages.java:22]: Thanks @cdrainCH and agree on the method name - have commented above.

File: src/main/java/uk/gov/companieshouse/overseasentitiesapi/validation/utils/DateValidators.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (mwestacott) [reply src/main/java/uk/gov/companieshouse/overseasentitiesapi/validation/utils/DateValidators.java:35]: Oh yes, sorry, I had trouble reading the logic and marrying it up to the message that gets output when check fails. All good.

**Pull request merged**